### ISSUE TITLE
Reveal Type and Hover Type inconsistency after type narrowing.

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [microsoft/pylance-release#2024]
## Environment data

-   Language Server version: v2021.10.3
-   OS and version: Mac 11.6
-   Python version (& distribution if applicable, e.g. Anaconda): Python 3.8

## Expected behaviour

reveal_type and hover type should always be same.

## Actual behaviour

Somehow hovering over bottom x gives me a type of Unknown, but the reveal type is int.

<img width="1792" alt="Screen Shot 2021-11-02 at 1 29 18 PM" src="https://user-images.githubusercontent.com/16809055/139924097-430d0f5e-e749-466f-9487-f534031ab528.png">

## Code Snippet / Additional information

```python
from typing import Optional


x: Optional[int] = None
for i in range(2):
    if i == 1:
        x = i

reveal_type(x)
assert x is not None
reveal_type(x)
```

I put this issue in pylance even though it's type checking related and maybe pyright is better because it's specifically hover type hint is off while reveal type hint is fine. This also seems weirdly specific to reveal_type as if I remove reveal_type here and make the code,

```python
from typing import Optional


x: Optional[int] = None
for i in range(2):
    if i == 1:
        x = i

reveal_type(x)
assert x is not None
x
```

then hovering over bottom x shows expected type again.

Main reason I find hover being wrong in reveal type scenario an issue is original program this comes from was longer and x's hover type becomes incorrect after more logic outside of reveal type too. This is was first place I could track to where type started doing weird things.

**Comment by erictraut** [comment]
Thanks for the bug report. I'm able to repro the problem. I've created a tracking bug in the pyright issue tracker.

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by heejaechang** [comment]
fixed in 2021.11.0
